### ISSUE TITLE
Update instructions re: how to get one's own app

### ISSUE THREAD

**Issue by jennybc** [r-lib/gmailr#143]
I used gmailr today for the time in a long time (definitely since the switch to gargle).

Overall, it was really smooth! I used the env var approach to convey the path to the JSON holding client ID and secret.

The process of getting one's own oauth app isn't reflected very well by the current README. I was able to do it because I have a lot of background knowledge, but it would be nice to update this. I might do this myself or perhaps someone else will do so.

**Comment by maelle** [comment]
I'd like to read these explanations as I suspect it'd solve #152 (or make me give up, if one needs to verify the app). :grin:

**Comment by thisisnic** [comment]
I just went through this process today to get `gmailr` working, and it wasn't too bad after a bit of googling and mild confusion.  If nobody is currently working on this, I'm happy to write up what I've learned and submit a PR updating the README?

**Comment by jtr13** [comment]
This is what worked for me today. It's an updated version of https://www.infoworld.com/article/3398701/how-to-send-email-from-r-and-gmail.html by Sharon Machlis. (Thank you Sharon!)  Happy to ask students to test and write up more formally somewhere if that's helpful.


1.  Go to <https://console.developers.google.com/apis/library>

2.  If you have more than one Google account, make sure you’re logged
    into the correct one–check the icon on the top right of the screen.
    It’s a good idea to only be logged into the account that you want to
    use.

3.  Click “Select a project”, “NEW PROJECT”, give it a project name,
    click “Create”. It’s ok if the Location fields says “No
    organization.” (Note: if you have other projects, creating a new
    project won’t necessarily move you into the new project, so be sure
    to choose “Select Project” after creating.)

4.  Search for “gmail”, choose Gmail API, and click “Enable”. If you
    don’t see a search option, click “APIs & Services” on the left and
    then “Library”. If it’s already enabled it will say “Manage”.

5.  Click “Create Credentials” then the following:

    -   Credential Type: Choose “Gmail API” for Select an API, choose
        “User data”, NEXT

    -   OAuth Consent Screen: Choose an App name (I don’t think this
        matters), and a support email and developer contact info (also
        don’t matter afaik), then SAVE AND CONTINUE.

    -   Scopes: Click SAVE AND CONTINUE

    -   OAuth Client ID: Choose “Desktop app” for the Application Type
        field. Choose any name or leave “Desktop client 1” for the name
        field. Click CREATE. Download the .json file. Click DONE.

6.  Choose the “OAuth consent screen” tab on the left. Click “PUBLISH
    APP” under “Testing”. Click “Confirm”.

Continue with Setup instructions in the  [README.md](https://github.com/r-lib/gmailr) starting with “Tell gmailr
where the JSON lives…”

**Comment by ThoDuyNguyen** [comment]
This approach seems not working now. Here is what I got (Client app was set as "Desktop App")

Code

```
options(
  gargle_oauth_cache = "REMOVE",
  gargle_verbosity = "debug",
  gargle_oauth_email = ""REMOVE""
)

gmailr::gm_auth_configure(path = "/path/to/json")

gmailr::gm_auth(email = "REMOVE")
``` 

Output 
```
trying `token_fetch()`
trying `credentials_service_account()`
Error caught by `token_fetch()`:
Argument 'txt' must be a JSON string, URL or file.
trying `credentials_external_account()`
aws.ec2metadata not installed; can't detect whether running on EC2 instance
trying `credentials_app_default()`
trying `credentials_gce()`
trying `credentials_byo_oauth()`
Error caught by `token_fetch()`:
inherits(token, "Token2.0") is not TRUE
trying `credentials_user_oauth2()`
Gargle2.0 initialize
adding "userinfo.email" scope
loading token from the cache
no matching token in the cache
initiating new token
```
The process halt. 

![image](https://user-images.githubusercontent.com/9877561/214546519-d9f03daf-9ff5-4db4-a1a9-27ae55eafe59.png)


![image](https://user-images.githubusercontent.com/9877561/214546265-12fcd9f4-4c3d-47f9-ba73-90b05583d784.png)

**Comment by maelle** [comment]
Not for GMail but trackdown has a nice article on setting up an app (for GMail one would "just" use other APIs/scopes) https://claudiozandonella.github.io/trackdown/articles/oauth-app-configuration.html

**Comment by jennybc** [comment]
Out-of-band auth has been partially shut down and that appears to play a role in what some people are seeing here (there is now a lot of different stuff reported in this thread). The latest gargle release has more details:

https://gargle.r-lib.org/news/index.html#gargle-130

To remind everyone, gmailr can't use the built-in tidyverse app, because we can't get those scopes approved.

So gmailr users often have to bring their own app and this has gotten noticeably more complicated for those who need out of band auth (OOB), i.e. where R/RStudio is running in the browser (vs. local computer). This is handled by what we now call pseudo-OOB. This requires a client of "web application" type.

These are the basic instruction for getting one's own app:

https://gargle.r-lib.org/articles/get-api-credentials.html#oauth-client-id-and-secret

If one is using a web application client and the pseudo-OOB flow, you'll also need to redirect to a static page that aids with token exchange. This is discussed in the later part of this thread:

https://github.com/r-lib/gargle/issues/202

The short version is that, at least for early experimentation, people can probably use the tidyverse page for exchange. Otherwise, you can serve your own somewhere, using example HTML provided in the gargle package.

If anyone pursues this, I'd love to hear from you, so we can form a more official stance on the callback page and improve the documentation.
